labs/b5.md

```diff
@@ -0,0 +1,364 @@
+
+TCP then begins transmitting data and if it successfully arrives on the other end of the connection then an ACK is issued. Therefore if data is lost, reordered, or corrupted, TCP is capable of recognizing this and sends a request for retransmission of any lost data. 
+
+TCP also has a procedure to close connections. We only consider a graceful termination here, abrupt terminations have a different procedure we will not go over. Let's assume machine A wants to close its connection to machine B.
+![alt text][termination]
+
+[termination]: http://intronetworks.cs.luc.edu/current/html/_images/tcp_closes.svg "TCP termination"
+A begins by sending a FIN. B must respond by sending a FIN and an ACK. If B only sends a ACK the connection persist and additional data can be sent until an FIN is sent. On the other hand B can also send just one packet with both FIN and ACK flags set, i.e. FIN+ACK if B is ready to close the connection and doesn't need to send additional data Once A has received a FIN and an ACK it sends one last ACK to signal the connection termination.
+
+UDP is stateless connectionless protocol. UDP focuses on sending messages in datagrams. Being stateless UDP also doesn't incur the overhead of the TCP handshake and termination. UDP also makes no guarantees about reliable transport so messages may be corrupted, arrive out of order, or not arrive at all. For this reason UDP is sometimes called Unreliable Datagram Protocol.
+
+While UDP makes no guarantees about reliable transport it doesn't suffer from the overhead of establishing and closing connections like in TCP. UDP is therefore ideal for usage cases where we just want to send packets quickly and losing a few of those isn't disastrous.
+
+Moreover, compared to TCP every single UDP send requires a UDP receive (recv) per datagram. While for TCP you pass a stream of data that is transparently split into some number of sends and the data stream can be reconstructed with a single recv call.
+
+## Ports
+[port list]: https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml
+
+Ports define a service endpoint, broadly speaking -- ports mark a point of traffic ingress and egress.  Ports are represented by a 16 bit number meaning thus ranging from 0 to 65535.  Ports from 0 to 1023 are well known ports, i.e. system ports.  Using these ports usually has a stricter requirement.  1024 to 49151 are registered ports. IANA maintains the official [list][port list] of well-known and registered ranges. The remaining ports from 49152 to 65535 are ephemeral ports which can be dynamically allocated for communication sessions on a per request basis.
```

Comment:
I'd say something like "IP addresses connect hosts to one another, ports connect processes to one another. Only one process(*) may be bound to a particular port associated with a particular IP address" and expound on that a bit, like what Shenker does in his lecture when he covers this stuff

Reply:
done and done